core/predicate_check.go

```diff
@@ -11,43 +11,45 @@ import (
 	"github.com/ava-labs/subnet-evm/precompile/precompileconfig"
 	predicateutils "github.com/ava-labs/subnet-evm/utils/predicate"
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
 )
 
-// CheckPredicates checks that all precompile predicates are satisfied within the current [predicateContext] for [tx]
-func CheckPredicates(rules params.Rules, predicateContext *precompileconfig.PredicateContext, tx *types.Transaction) error {
-	// Short circuit early if there are no precompile predicates to verify
-	if len(rules.Predicates) == 0 {
-		return nil
-	}
-
+// CheckPredicates verifies the predicates of [tx] and returns the result. Returning an error invalidates the block.
+func CheckPredicates(rules params.Rules, predicateContext *precompileconfig.PredicateContext, tx *types.Transaction) (map[common.Address][]byte, error) {
 	// Check that the transaction can cover its IntrinsicGas (including the gas required by the predicate) before
 	// verifying the predicate.
 	intrinsicGas, err := IntrinsicGas(tx.Data(), tx.AccessList(), tx.To() == nil, rules)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	if tx.Gas() < intrinsicGas {
-		return fmt.Errorf("insufficient gas for predicate verification (%d) < intrinsic gas (%d)", tx.Gas(), intrinsicGas)
+		return nil, fmt.Errorf("%w for predicate verification (%d) < intrinsic gas (%d)", ErrIntrinsicGas, tx.Gas(), intrinsicGas)
 	}
 
-	// Track addresses that we've performed a predicate check for
-	precompileAddressChecks := make(map[common.Address]struct{})
+	predicateResults := make(map[common.Address][]byte)
+	// Short circuit early if there are no precompile predicates to verify
+	if len(rules.Predicates) == 0 {
+		return predicateResults, nil
+	}
+	predicateArguments := make(map[common.Address][][]byte)
 	for _, accessTuple := range tx.AccessList() {
 		address := accessTuple.Address
-		predicater, ok := rules.Predicates[address]
+		_, ok := rules.Predicates[address]
 		if !ok {
 			continue
 		}
-		// Return an error if we've already checked a predicate for this address
-		if _, ok := precompileAddressChecks[address]; ok {
-			return fmt.Errorf("predicate %s failed verification for tx %s: specified %s in access list multiple times", address, tx.Hash(), address)
-		}
-		precompileAddressChecks[address] = struct{}{}
-		predicateBytes := predicateutils.HashSliceToBytes(accessTuple.StorageKeys)
-		if err := predicater.VerifyPredicate(predicateContext, predicateBytes); err != nil {
-			return fmt.Errorf("predicate %s failed verification for tx %s: %w", address, tx.Hash(), err)
-		}
+
+		predicateArguments[address] = append(predicateArguments[address], predicateutils.HashSliceToBytes(accessTuple.StorageKeys))
```

Comment:
This was so that the predicate can take in all of the byte slices for the transaction in one call rather than doing it one at a time and return the bit set as a byte slice.

Alternatively, we could switch back to `VerifyPredicate` taking in a single value and `CheckPredicates` assembling the bit set from those results. This would make the code a bit simpler at the cost of constraining the precompile predicates to return a true/false value instead of an arbitrary byte array.

Reply:
I feel like using a single value for `VerifyPredicate` could be simpler + cleaner (and costs less loops). It would be easier for 3rd party implementations as well.

Reply:
Could we make this a follow-on PR to try out the alternative?

This also means that we need to track the indices used by each precompile address to create that bitset, so I'm not sure if it will be cleaner.

Reply:
Created https://github.com/ava-labs/subnet-evm/issues/844